### PR TITLE
Feat/pytorch support

### PR DESCRIPTION
**Pull request by esherman-credo**
## Describe your changes
Add support for pytorch models.
- Modify base_data to accept pytorch dataloader objects --> this could potentially work out of box for keras dataloaders, but unclear at the moment
- Modify classification model to support PyTorch models with HEAVY constraints -> user must specify some extra attributes in their custom-defined model class

## Issue ticket number and link
https://credo-ai.atlassian.net/browse/DSP-475

## Known outstanding issues that are not fully accounted for
- Need to confirm data validation is what we want it to be
- Comments

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have built basic tests for new functionality (particularly new evaluators)
- [ ] If new libraries have been added, I have checked that [readthedocs](https://readthedocs.org/projects/credoai-lens/) API documentation is constructed correctly
- [x] Will this be part of a major product update? If yes, please write one phrase about this update.
    - This adds support for PyTorch. It is very limited at this time (just classification models with essentially sequential architectures). Allows us to "play in" the PyTorch ecosystem.

## Extra-mile Checklist
- [ ] I have thought expansively about edge cases and written tests for them

### CODE REVIEW

File: credoai/artifacts/model/regression_model.py [hunk 1]
```diff
@@ -1,6 +1,15 @@
 """Model artifact wrapping any regression model"""
 from .base_model import Model
 
+from utils.model_utils import reg_handle_torch
+
+try:
```
Comment (fabrizio-credo) [inline credoai/artifacts/model/regression_model.py:6]: As it is now this will be triggered anytime a user is using the artifact and has no `torch` installed. Can it be made specific to only attempts to handle torch models without the `torch` library?

File: credoai/utils/model_utils.py [hunk 2]
```diff
@@ -125,3 +170,165 @@ def validate_dummy(model_like, _):
...
+    ):
+        predict_proba_obj = pred_func
+    else:
+        pass
+        # predict_proba is not valid (for now) --> this would correspond to multi-dimensional outputs or something similarly weird
+
+    return predict_obj, predict_proba_obj, clf_type
+
+
+def clf_handle_torch(model_like):
```
Comment (fabrizio-credo) [inline credoai/utils/model_utils.py:223]: I like how you separated handle functions. Can you add a docstring that gives the gist of what is going on?


```
### This is what Claude suggests 😛 
Creates a standardized interface for making inferences from a PyTorch classification model with different activations and output types.

- It checks that the model passed in has a forward method for making predictions. If not, it raises an error.
- It gets the output activation used by the model, either "sigmoid" or "softmax". If an invalid activation is specified, it defaults to "sigmoid".
- It performs a dummy prediction with the model to get the shape of the output. It uses this to determine if the model is a binary or multi-class classifier.
- It creates a predict function that performs inferences on either a single input or a batch of inputs. For sigmoid activation, it returns rounded predictions. For softmax, it returns the predicted class label.
- It creates a predict_proba function to return raw prediction probabilities. For sigmoid, it returns the logistic prediction. For softmax, it returns the probability of the positive class for binary classification or the full probability distribution for multi-class classification.
- It returns the predict and predict_proba functions along with a string specifying the classifier type (binary or multi-class).
- These returned functions can then be used to easily make predictions and get probabilities from the PyTorch model. The functions handle the differences between sigmoid and softmax activations and between binary and multi-class classification models.
"""
```

**Review by fabrizio-credo — CHANGES_REQUESTED** [review]
All minor stuff, except the library import ones which for me are breaking.

Excellent work!

**Review by IanAtCredoAI — APPROVED** [review]
Looked through everything. The main change I made was making a PyTorch data class so that we don't make the Data class overly ugly.

Heroic work with the clf/reg handlers in model_utils. It's not the prettiest, but I was able to read through the web easily.

**Comment by IanAtCredo** [comment]
Removed approval as I am building an example notebook. In that process it seems that my pytorch model is not being appropriately recognized (framework isn't being set correctly)

**Comment by IanAtCredo** [comment]
Issues still outstanding:
* clf_torch_handle needs to define prediction functions that make use of the device used for training
* `output_activation` is janky. The fact that it needs to be specified isn't great. In addition, the implementation applies sigmoid/softmax to the output (as if the output was a linear layer). That's not necessarily the case.
